src/sources/syslog.rs

```diff
@@ -219,59 +223,108 @@ impl SyslogDecoder {
-                    .map_err(|_| {
-                        LinesCodecError::Io(io::Error::new(
+        let space_pos = src.iter().position(|&b| b == b' ');
+
+        // If we are discarding, discard to the next newline.
+        let newline_pos = src.iter().position(|&b| b == b'\n');
+
+        match (self.is_discarding, newline_pos, space_pos) {
+            (true, Some(offset), _) => {
+                // When discarding we keep discarding to the next newline.
+                src.advance(offset + 1);
+                self.is_discarding = false;
+                self.in_octet = false;
+                Err(LinesCodecError::Io(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Frame length limit exceeded",
+                )))
+            }
+            (true, None, _) => {
+                // There is no newline in this frame. Advance as far as we can to
+                // discard the entire frame.
+                src.advance(src.len());
+                Ok(None)
+            }
+            (false, _, Some(space_pos)) => {
+                let len: usize = match std::str::from_utf8(&src[..space_pos])
+                    .map_err(|_| ())
+                    .and_then(|num| num.parse().map_err(|_| ()))
+                {
+                    Ok(len) => len,
+                    Err(_) => {
+                        // Advance the buffer past the erroneous bytes
+                        // to prevent us getting stuck in an infinite loop.
+                        src.advance(space_pos + 1);
+                        self.in_octet = false;
+                        return Err(LinesCodecError::Io(io::Error::new(
                             io::ErrorKind::InvalidData,
-                            "Unable to decode message as UTF8",
-                        ))
-                    })?
-                    .to_string();
-                src.advance(to);
-                Ok(Some(s))
-            } else {
+                            "Unable to decode message len as number",
+                        )));
+                    }
+                };
+
+                let from = space_pos + 1;
+                let to = from + len;
+
+                if let Some(msg) = src.get(from..to) {
+                    let s = match std::str::from_utf8(msg) {
+                        Ok(s) => s.to_string(),
+                        Err(_) => {
+                            // Advance the buffer past the erroneous bytes
+                            // to prevent us getting stuck in an infinite loop.
+                            src.advance(to);
+                            self.in_octet = false;
+                            return Err(LinesCodecError::Io(io::Error::new(
+                                io::ErrorKind::InvalidData,
+                                "Unable to decode message as UTF8",
+                            )));
+                        }
+                    };
+
+                    src.advance(to);
```

Comment:
It seems like advancing the buffer can be done unconditionally here:

```suggestion
                    src.advance(to);

                    let s = match std::str::from_utf8(msg) {
                        Ok(s) => s.to_string(),
                        Err(_) => {
                            self.in_octet = false;
                            return Err(LinesCodecError::Io(io::Error::new(
                                io::ErrorKind::InvalidData,
                                "Unable to decode message as UTF8",
                            )));
                        }
                    };
```

Reply:
It would appear that way, but unfortunately this gives us a number of lifetime issues since `src.get` returns a reference which means that `src.advance` can't mutably borrow until we have finished the `match`.